dbms/src/Functions/LeastGreatest.h

```diff
@@ -120,5 +122,289 @@ class FunctionVectorizedLeastGreatest : public IFunction
+        const ColumnString::Chars_t & b_data,
+        const ColumnString::Offsets & b_offsets,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets,
+        size_t i)
+    {
+        size_t a_size;
+        size_t b_size;
+
+        if (i == 0)
+        {
+            a_size = a_offsets[0] - 1;
+            b_size = b_offsets[0] - 1;
+            processImpl(collator, a_size, b_size, &a_data[0], &b_data[0], c_data, c_offsets);
+        }
+        else
+        {
+            a_size = a_offsets[i] - a_offsets[i - 1] - 1;
+            b_size = b_offsets[i] - b_offsets[i - 1] - 1;
+            processImpl(collator, a_size, b_size, &a_data[a_offsets[i - 1]], &b_data[b_offsets[i - 1]], c_data, c_offsets);
+        }
+    }
+
+    // string_constant
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const ColumnString::Chars_t & a_data,
+        const ColumnString::Offsets & a_offsets,
+        const String & b,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets,
+        size_t i)
+    {
+        const auto * b_data = reinterpret_cast<const unsigned char *>(b.data());
+        ColumnString::Offset b_size = b.size();
+        size_t a_size;
+        if (i == 0)
+        {
+            a_size = a_offsets[0] - 1;
+            processImpl(collator, a_size, b_size, &a_data[0], &b_data[0], c_data, c_offsets);
+        }
+        else
+        {
+            a_size = a_offsets[i] - a_offsets[i - 1] - 1;
+            processImpl(collator, a_size, b_size, &a_data[a_offsets[i - 1]], &b_data[0], c_data, c_offsets);
+        }
+    }
+
+    // constant_constant
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const std::string & a,
+        const std::string & b,
+        std::string & c)
+    {
+        int res = 0;
+        if (!collator)
+        {
+            res = memcmp(a.data(), b.data(), std::min(a.size(), b.size()));
+        }
+        else
+        {
+            res = collator->compare(reinterpret_cast<const char *>(a.data()), a.size(), reinterpret_cast<const char *>(b.data()), b.size());
+        }
+        if (res < 0)
+            c = a;
+        else if (res == 0)
+        {
+            size_t size = std::min(a.size(), b.size());
+            c = b.substr(0, size);
+        }
+        else
+            c = b;
+    }
+};
+
+struct StringOperationImpl
+{
+    static void NO_INLINE stringVectorStringVector(
+        const ColumnString::Chars_t & a_data,
+        const ColumnString::Offsets & a_offsets,
+        const ColumnString::Chars_t & b_data,
+        const ColumnString::Offsets & b_offsets,
+        const TiDB::TiDBCollatorPtr & collator,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets)
+    {
+        size_t size = a_offsets.size();
+        c_data.reserve(std::max(a_data.size(), b_data.size()));
+        for (size_t i = 0; i < size; ++i)
```

Comment:
I think we should put the `i == 0` case out of the for-loop so that we can avoid `if` calse in `process` and make compiler tend to generate vectorization codes.